src/passes/PostAssemblyScript.cpp

```diff
@@ -0,0 +1,530 @@
+static Name ALLOC = Name("~lib/rt/tlsf/__alloc");
+static Name ALLOCARRAY = Name("~lib/rt/__allocArray");
+
+// A variant of LocalGraph that considers only assignments when computing
+// influences.
+//
+// This allows us to find locals aliasing a retain, while ignoring other
+// influences.
+//
+// For example, we are interested in
+//
+//   var a = __retain(X)
+//   var b = a;
+//   __release(b); // releases X
+//
+// but not in
+//
+//  var a = __retain(X);
+//  var b = someFunction(a);
+//  __release(b);
+//  return a;
+//
+// since the latter releases not 'X' but the reference returned by the call,
+// which is usually something else.
+struct AliasGraph : LocalGraph {
+  AliasGraph(Function* func) : LocalGraph(func) {}
+  void computeInfluences() {
+    for (auto& pair : locations) {
+      auto* curr = pair.first;
+      if (auto* set = curr->dynCast<LocalSet>()) {
+        if (auto* get = set->value->dynCast<LocalGet>()) {
+          getInfluences[get].insert(set);
+        }
+      } else {
+        auto* get = curr->cast<LocalGet>();
+        for (auto* set : getSetses[get]) {
+          setInfluences[set].insert(get);
+        }
+      }
+    }
+  }
+};
+
+// A pass that eliminates redundant retain and release calls.
+//
+// Does a cheap traversal first, remembering ARC-style patterns, and goes all-in
+// only if it finds any.
+//
+// This is based on the assumption that the compiler is not allowed to emit
+// unbalanced retains or releases, except if a value is returned or otherwise
+// escapes in one branch. In turn, we do not have to deal with control
+// structures but can instead look for escapes reached (by any alias) using
+// AliasGraph.
+//
+// For example, in code like
+//
+//   var a = __retain(X);
+//   if (cond) {
+//     return a;
+//   }
+//   __release(a);
+//  return null;
+//
+// we cannot eliminate the retain/release pair because the implementation
+// dictates that returned references must remain retained for the caller since
+// dropping to RC=0 on the boundary would prematurely free the object.
+//
+// Typical patterns this recognizes are
+//
+//   var a = __retain(X);
+//   __release(a); // simple pair
+//
+//   var a = __retain(X);
+//   if (cond) {
+//     __release(a); // non-escaping...
+//   } else {
+//     __release(a); // ...balanced pair
+//   }
+//
+//   var a;
+//   if (cond) {
+//     a = __retain(X);
+//   } else {
+//     a = __retain(Y);
+//   }
+//   __release(a); // balanced retain
+//
+// including technically invalid patterns assumed to be not present in compiler
+// output, like:
+//
+//   var b = __retain(a);
+//   if (cond) {
+//     __release(b); // unbalanced release
+//   }
+//
+// To detect the latter, we'd have to follow control structures around, which
+// we don't do since it isn't neccessary / to keep the amount of work minimal.
+struct OptimizeARC : public WalkerPass<PostWalker<OptimizeARC>> {
+
+#ifndef POST_ASSEMBLYSCRIPT_DEBUG
```

Comment:
This ifdef is risky, as the semantics are different. A function parallel pass gets a new instance per function; a normal pass does not. In particular, the `balancedRetainsCache` object will accumulate data in the latter case, I think...

Note that if you do `BINARYEN_PASS_DEBUG=1` in the env then I think passes run with a single core, which is nice for debugging.

Reply:
Good to know, thanks :) Removed the ifdef.